skfda/preprocessing/dim_reduction/projection/_fpca.py

```diff
@@ -1,33 +1,532 @@
+    fdata2pc of the library fda.usc in R, and the referenced paper.
+
+    Args:
+        sample_points: the points of discretization of the data matrix.
+        penalty (array_like): coefficients representing the differential
+            operator used in the computation of the penalty matrix. For example,
+            the array (1, 0, 1) means :math:`1 + D^{2}`
     Returns:
-        tuple: (scores, principal directions, eigenvalues)
+        (array_like): the penalty matrix used to regularize the components
+
+    References.
+        [1] Nicole Krämer, Anne-Laure Boulesteix, and Gerhard Tutz. Penalized
+        partial least squares with applications to b-spline transformations
+        and functional data. Chemometrics and Intelligent Laboratory Systems,
+        94:60–69, 11 2008.
 
     """
-    fdatagrid = fdatagrid - mean(fdatagrid)  # centers the data
-    # singular value decomposition
-    u, s, v = np.linalg.svd(fdatagrid.data_matrix)
-    principal_directions = v.T  # obtain the eigenvectors matrix
-    eigenvalues = (np.diag(s) ** 2) / (fdatagrid.n_samples - 1)
-    scores = u @ s  # functional principal scores
-
-    return scores, principal_directions, eigenvalues
+    penalty = np.array(penalty)
+    n_points = len(sample_points)
+    penalty_matrix = np.zeros((n_points, n_points))
+    if np.sum(penalty) != 0:
+        # independent term
+        penalty_matrix = penalty_matrix + penalty[0] * np.diag(
+            np.ones(n_points))
+        if len(penalty) > 1:
+            # for each term of the differential operator, we compute the penalty
+            # matrix of that order and then add it to the final penalty matrix
+            aux_penalty_1 = _auxiliary_penalty_matrix(sample_points)
+            aux_penalty_2 = _auxiliary_penalty_matrix(sample_points)
+            for i in range(1, len(penalty)):
+                if i > 1:
+                    aux_penalty_1 = (aux_penalty_2[:(n_points - i),
+                                     :(n_points - i + 1)]
+                                     @ aux_penalty_1)
+                # applying the differential operator, as in each step the
+                # derivative degree increases by 1.
+                penalty_matrix = (penalty_matrix +
+                                  penalty[i] * (np.transpose(
+                                    aux_penalty_1) @ aux_penalty_1))
+    return penalty_matrix
+
+
+class FPCAGrid(FPCA):
+    """Funcional principal component analysis for functional data represented
+    in discretized form.
+
+    Parameters:
+        n_components (int): number of principal components to obtain from
+            functional principal component analysis. Defaults to 3.
+        centering (bool): if True then calculate the mean of the functional data
+            object and center the data first. Defaults to True. If True the
+            passed FDataBasis object is modified.
+        weights (numpy.array Union callable): the weights vector used for
```

Comment:
```suggestion
        weights (numpy.array or callable): the weights vector used for
```
(or use the Union notation properly).